### PR TITLE
Fork tree prune assumptions removal v2

### PR DESCRIPTION
**Pull request by davxy**
Superseeds https://github.com/paritytech/substrate/pull/12778

---

Fork tree prune method was **assuming** that the user's `predicate` closure makes the `find_node_index_where` to **NOT** return the tree node immediately preceding the `hash` parameter (according to `is_descendent_of` closure).

Up to date the only users of this structure are GRANDPA and BABE to keep a tree of epoch changes announcements AND according to their `predicate`, the assumption is satisfied.

But this is a time-bomb

If at some point the struct is used in a context where the predicate doesn't satisfy the assumption, the tree would break and we loose part of the tree due to how `prune` logic is implemented.
(indeed this assumption was the cause of https://github.com/paritytech/substrate/pull/12758 bug)

This PR removes the assumption and makes the tree more resilient and independent for any use case.

---

- [x] Burn-in test
- [x] Test in live network

### CODE REVIEW

File: utils/fork-tree/src/lib.rs [hunk 1]
```diff
@@ -32,7 +32,7 @@ pub enum Error<E> {
 	UnfinalizedAncestor,
 	/// Imported or finalized node that is an ancestor of previously finalized node.
 	Revert,
-	/// Error throw by client when checking for node ancestry.
+	/// Error thrown by user when checking for node ancestry.
 	Client(E),
```
Comment (andresilva) [inline utils/fork-tree/src/lib.rs:36]: Maybe rename this to `Ancestry(E)`? Or something else if someone has a better suggestion.

File: utils/fork-tree/src/lib.rs [hunk 2]
```diff
@@ -99,12 +95,14 @@ where
...
 	N: Ord,
 {
-	/// Create a new empty tree.
+	/// Create a new empty tree instance.
 	pub fn new() -> ForkTree<H, N, V> {
 		ForkTree { roots: Vec::new(), best_finalized_number: None }
 	}
 
-	/// Rebalance the tree, i.e. sort child nodes by max branch depth (decreasing).
+	/// Rebalance the tree
```
Comment (andresilva) [inline utils/fork-tree/src/lib.rs:103]:
```suggestion
	/// Rebalance the tree.
```

File: utils/fork-tree/src/lib.rs [hunk 3]
```diff
@@ -371,42 +376,59 @@ where
...
+			for (idx, child) in curr.children.iter().enumerate() {
+				if child.number < *number && is_descendent_of(&child.hash, hash)? {
+					maybe_ancestor_idx = Some(idx);
+					break
+				}
+			}
+			let Some(ancestor_idx) = maybe_ancestor_idx else {
+				// Now we are positioned just above block identified by `hash`
+				break
+			};
```
Comment (andresilva) [inline utils/fork-tree/src/lib.rs:410]: I guess something went wrong here.

File: client/consensus/epochs/src/lib.rs [hunk 4]
```diff
@@ -1039,7 +1039,7 @@ mod tests {
 
 		let mut nodes: Vec<_> = epoch_changes.tree().iter().map(|(h, _, _)| h).collect();
 		nodes.sort();
-		assert_eq!(nodes, vec![b"A", b"B", b"C", b"E", b"F", b"G"]);
+		assert_eq!(nodes, vec![b"A", b"B", b"C", b"F", b"G"]);
```
Comment (andresilva) [inline client/consensus/epochs/src/lib.rs:1042]: Nice.

File: utils/fork-tree/src/lib.rs [hunk 5]
```diff
@@ -1332,6 +1354,77 @@ mod test {
...
+		);
+
+		assert_eq!(
+			removed.map(|(hash, _, _)| hash).collect::<Vec<_>>(),
+			vec!["O", "I", "G", "A", "B", "C", "D", "E", "J", "K"]
+		);
+	}
+
+	#[test]
+	fn prune_works5() {
```
Comment (andresilva) [inline utils/fork-tree/src/lib.rs:1413]: Are we testing anything different here?

**Review by andresilva — APPROVED** [review]
lgtm, good catch

File: utils/fork-tree/src/lib.rs (hunk 3 above)
Comment (davxy) [reply utils/fork-tree/src/lib.rs:410]: bleeding edge stuff 😆

**Review by skunert — APPROVED** [review]
LGTM, nice readability improvements over the last PR

**Review by melekes — COMMENTED** [review]
lgtm

**Pull request merged**